lib/logstorage/storage.go

```diff
@@ -302,7 +314,7 @@ func (s *Storage) runRetentionWatcher() {
 }
 
 func (s *Storage) watchRetention() {
-	ticker := time.NewTicker(time.Hour)
+	ticker := time.NewTicker(time.Duration(s.retentionWatchInterval) * time.Hour)
```

Comment:
VictoriaLogs deletes per-day partitions outside the configured retention once per hour. This means that queries can return logs with timestamps for up to one day outside the configured retention. This behaviour can be unexpected by users, and it would be great to fix it.

This change allows configuring the interval for checking whether some per-day partitions are ready to be deleted, and the minimum allowed interval is one hour. This change doesn't solve the unexpected behaviour when logs outside the configured retention can be returned from queries.

It would be great to see real production use case for this change though.

Reply:
Thank you for your reply.

> This change doesn't solve the unexpected behaviour when logs outside the configured retention can be returned from queries.

Yes, I agree with this.

But this PR is not used to fix this issue, in our use case, we hope to reduce the frequency of querying expired partitions, so as to reduce `s.partitionsLock` conflicts, and we even want this trigger to be executed only in the early morning every day.

Although reducing the frequency does not significantly improve performance, I think giving users another choice is not a bad thing.

---

> This behaviour can be unexpected by users, and it would be great to fix it.

By the way, although this issue has little impact on our use case, we can additionally judge whether the partition has expired when querying the data, and then delete the partition, similar to `Redis`.

Reply:
> But this PR is not used to fix this issue, in our use case, we hope to reduce the frequency of querying expired partitions, so as to reduce s.partitionsLock conflicts, and we even want this trigger to be executed only in the early morning every day.

> Although reducing the frequency does not significantly improve performance, I think giving users another choice is not a bad thing.

The `partitionsLock` is acquired at every request for data ingestion and querying. The frequency of these requests is at least thousands times higher than the frequency of the check for outdated partitions, which is performed once per hour now. There is no any practical sense in reducing the frequency for the outdated partitions' check further.